Terminator strings for generate()

Rocketknight1 opened the pull request:
`generate()` stops when it encounters `eos_token_id`, but there are various circumstances when we want it to stop for other tokens too. The ideal situation would be to allow a set of strings that halts generation, and then include this information with the model, so model authors can set e.g. custom tokens like `<|im_end|>` as halting strings, even when those strings don't have a special token.

The problem with stopping for specific strings rather than tokens is that a string can be tokenized in many different ways, and the tokens that contain a string may also have overhangs on either end: `["?><", "|", "im_", "end", "|", ">>"]`. Since we have to check after each token generated by the model, we want to avoid detokenization and string comparisons, as this will cause a lot of slowdown and prevent us from compiling the generation loop.

This PR adds a `StoppingCriteria` for stop strings. It takes some time to preprocess the stop strings and the tokenizer vocabulary together and builds an embedding matrix containing the information it needs about which tokens can construct each stop string, but once that's done the entire generation-time check can be performed with only tensor operations and static, known shapes.

fixes #28801

Rocketknight1 committed:
stash commit (will discard all of this)

Rocketknight1 committed:
stash commit

Rocketknight1 committed:
First commit - needs a lot of testing!

Rocketknight1 committed:
Add a test

Rocketknight1 committed:
Fix imports and make the tests actually test something

Rocketknight1 committed:
Tests pass!

Rocketknight1 committed:
Rearrange test

Rocketknight1 committed:
Add comments (but it's still a bit confusing)

Rocketknight1 committed:
Stop storing the tokenizer

Rocketknight1 committed:
Comment fixup

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28932). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Rocketknight1 committed:
Fix for input_ids with a single sequence

Rocketknight1 committed:
Update tests to test single sequences

Rocketknight1 committed:
make fixup

zucchini-nlp commented:
@Rocketknight1 , hey! I liked the feature, a very useful one I think. Just a couple questions, since I am not sure what was the intended behavior initially

zucchini-nlp commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,167 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+
+        # Now we concatenate the match booleans across all strings and check if any are True
+        string_matches = torch.cat(string_matches, dim=0)
+        return torch.any(string_matches).item()
```
just being curious, so the generation stops for all batches when at least one has stop string?

Rocketknight1 replied:
Yes, this is correct! I think this is the desired behaviour for all stopping conditions (cc @gante to confirm)

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,167 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+                matching_positions = []
+                possible_end_lengths = []
+                for i in range(1 - len(token), len(stop_string)):
+                    tok = token[::-1].replace("▁", " ")
```
Ah, this is a really good point! I forgot that different tokenizers have different ways of encoding spaces. I should look for a correct general method to convert tokens to strings instead of writing all of the different special cases

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,167 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+            # Note that we're actually tracking one cumsum() for the list for each possible start overhang length
+            later_match = torch.isin(cumsum[:, :-1], valid_positions)
+            # The match vector is a boolean vector that indicates which positions have valid tokens
+            match = torch.cat([initial_match, later_match], dim=1)
```
I didn't spot this in testing! Can you give me the sample code that reproduces the issue?

zucchini-nlp replied:
You can use a stop string "class" or "end", both will give 2-dim `starts` tensor

Rocketknight1 replied:
I tried this and found a couple of related issues, which should be fixed now. I tested "stop", "end" and "s" as stop strings and it seemed to work properly. Can you let me know if you're still seeing the problem, and if so, can you paste me some code to reproduce it?

zucchini-nlp replied:
Thanks for iterating! I pulled the new changes, but the problem persists. The minimal code is same as the testing code, but with "end" added to stop strings

```
def test_stop_string_criteria():
        true_strings = ["<|im_start|><|im_end|>", "<|im_start|><|im_end|<|im_end|>", ">><|im_start|>>stop", "stop"]
        false_strings = [
            "<|im_start|><|im_end|",
            "<|im_start|><|im_end|<|im_end|",
            "<|im_end|><|im_start|>",
            "<|im_end|<>stop<|im_end|",
        ]
        too_short_strings = ["<|im_end|", "|im_end|>", "s", "end"]

        # Use a tokenizer that won't actually have special tokens for these
        tokenizer = AutoTokenizer.from_pretrained("openai-community/gpt2")
        tokenizer.pad_token_id = tokenizer.eos_token_id
        tokenizer.padding_side = "left"
        true_input_ids = tokenizer(true_strings, return_tensors="pt", padding="longest", add_special_tokens=False)
        false_input_ids = tokenizer(false_strings, return_tensors="pt", padding="longest", add_special_tokens=False)
        too_short_input_ids = tokenizer(
            too_short_strings, return_tensors="pt", padding="longest", add_special_tokens=False
        )
        scores = None

       # here 'end' is added to the list
        criteria = StopStringCriteria(tokenizer=tokenizer, stop_strings=["<|im_end|>", "stop", "end"])
        for i in range(len(true_strings)):
            assert(criteria(true_input_ids["input_ids"][i : i + 1], scores))
        for i in range(len(false_strings)):
            assert not (criteria(false_input_ids["input_ids"][i : i + 1], scores))
        for i in range(len(too_short_strings)):
            assert not (criteria(too_short_input_ids["input_ids"][i : i + 1], scores))
```

Rocketknight1 replied:
@zucchini-nlp found it! I was making quite a bad mistake with the array shapes. Should be fixed now!

Rocketknight1 committed:
Fix incorrect use of isin()

Rocketknight1 committed:
Expand tests to catch more cases

Rocketknight1 committed:
Expand tests to catch more cases

Rocketknight1 committed:
make fixup

Rocketknight1 committed:
Fix length calculation and update tests

Rocketknight1 committed:
Handle Ġ as a space replacement too

amyeroberts commented:
Nice! I'll let @gante review first to confirm it's all aligned with the current logic processors. 

Just skimming my main comment is that we need tests for the criterion's methods, in particular `get_matching_positions`.

gante commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,172 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+                    strings_to_end_lengths[stop_string][token] = possible_end_lengths
+        return strings_to_valid_positions, strings_to_end_lengths
+
+    def create_embedding_vecs(self):
```
```suggestion
    def create_embedding_vecs(self) -> Dict[str, torch.tensor]:
```

gante commented on tests/generation/test_stopping_criteria.py:
```diff
@@ -107,3 +108,57 @@ def test_validate_stopping_criteria(self):
         self.assertEqual(len(stopping_criteria), 1)
+
+    def test_stop_string_criteria(self):
+        true_strings = [
```
nit: I would add `stop_strings = ["<|im_end|>", "stop", "end"]` before the test cases, so we better understand the categorization of the test strings. I wondered whether the correct stop strings came from when reading this function top to bottom :D

gante commented on tests/generation/test_stopping_criteria.py:
```diff
@@ -107,3 +108,57 @@ def test_validate_stopping_criteria(self):
+            "<|im_start|><|im_end|",
+            "<|im_start|><|im_end|<|im_end|",
+            "<|im_end|><|im_start|>",
+            "<|im_end|<>stop<|im_end|",
```
```suggestion
            "<|im_end|<>stop<|im_end|",
            "en d",
            "eNd",
```
just in case we do some space/capitalization shenanigans in the future 🙃

Rocketknight1 replied:
Done!

gante reviewed (commented):
Very cool PR 🔥 

Extra request: let's add `stop_strings` to the `GenerationConfig` and, if it is set at generation time, let's add this stopping criteria in `_get_stopping_criteria` (pro tip: we can instantiate the tokenizer from the model repo attribute in the model instance, so we don't need to pass the tokenizer to `generate` :D). That way, users can:
1. do `model.generate(..., stop_strings=["foo", "bar"])`, as opposed to `model.generate(..., stopping_criteria=StoppingCriteriaList(...))`, which is more user-friendly
2. store their model's `stop_strings` in the generation config 💛

Rocketknight1 commented:
@amyeroberts those are purely internal methods - maybe I should just mark them as private with a leading `_` instead?

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,172 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+        self.embedding_vecs = self.create_embedding_vecs()
+
+    @staticmethod
+    def get_matching_positions(vocab, stop_strings):
```
This is correct, I think!

Rocketknight1 committed:
Update src/transformers/generation/stopping_criteria.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

Rocketknight1 committed:
Add optimizations from Joao's suggestion

Rocketknight1 committed:
Remove TODO

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,172 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+    @add_start_docstrings(STOPPING_CRITERIA_INPUTS_DOCSTRING)
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
+        # TODO Joao - I'm not using the scores at all and just checking the most recent tokens in input_ids
+        #      Is this correct? Should I be sampling from scores?
```
Got it! Removed that TODO

Rocketknight1 committed:
Update src/transformers/generation/stopping_criteria.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

Rocketknight1 committed:
Update tests/generation/test_stopping_criteria.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

Rocketknight1 committed:
make fixup

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,172 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+        stop_strings (`Union[str, List[str]]`):
+            A list of strings that should end generation. If a string is passed, it will be treated like a
+            list with a single element.
+    """
```
I wasn't sure which examples you mean! Can you link me to the code?

gante replied:
e.g.
https://github.com/huggingface/transformers/blob/6d1f545665ac66420af9f6702d891a30c5d070ea/src/transformers/generation/logits_process.py#L113

Rocketknight1 replied:
Ah, got it! Let me see if I can incorporate this into generate + the `generation_config` first, and then write those

Rocketknight1 replied:
Done! (With apologies for the delay when I got pulled away to work on other stuff)

The example now lives [here](https://github.com/huggingface/transformers/pull/28932/files#diff-6e63ae0764aa864afd5bae6d512677b99b5240cb98cb210190482bdbb6a85906R263-R285). I tested it and it works well!

Rocketknight1 committed:
Rename some variables and remove some debugging clauses for clarity

Rocketknight1 committed:
Add tests for the sub-methods

Rocketknight1 commented:
@amyeroberts tests for the sub-methods are in!

Rocketknight1 committed:
Clarify one test slightly

Rocketknight1 committed:
Add stop_strings to GenerationConfig

Rocketknight1 committed:
generate() supports stop_string arg, asks for tokenizer if not provided

Rocketknight1 committed:
make fixup

Rocketknight1 committed:
Cleanup code and rename variables for clarity

Rocketknight1 committed:
Update tokenizer error

Rocketknight1 committed:
Update tokenizer passing, handle generation on GPU

Rocketknight1 committed:
Slightly more explanation cleanup

Rocketknight1 committed:
More comment cleanup

Rocketknight1 committed:
Factor out the token cleanup so it's more obvious what we're doing, and we can change it later

Rocketknight1 committed:
Careful with that cleanup!

Rocketknight1 committed:
Cleanup + optimizations to _get_matching_positions

Rocketknight1 committed:
More minor performance tweaks

Rocketknight1 committed:
Implement caching and eliminate some expensive ops (startup time: 200ms -> 9ms)

Rocketknight1 committed:
Remove the pin_memory call

Rocketknight1 committed:
Parallelize across all stop strings!

Rocketknight1 committed:
Quick fix for tensor devices

Rocketknight1 committed:
Update embeddings test for the new format

Rocketknight1 committed:
Fix test imports

Rocketknight1 commented:
Quick update here: I refactored the initialization and added a small cache in case users repeatedly call `generate()`. Initialization time is down from 200ms to ~10ms in local testing. I also rewrote the core call so that all stop strings are tested in parallel - this results in a big speedup when we're testing more than 1 stop string.

We may still end up going with @amyeroberts' tokenizer-decode solution after profiling, but I wanted to make sure this method didn't have any obvious performance issues first :fist_raised:. I'll do some testing on Monday!

(also I still have a deep phobia of graph breaks from my XLA era)

Rocketknight1 commented:
:relieved:
![image](https://github.com/huggingface/transformers/assets/12866554/08d6e4fb-8665-4083-952d-a721c18da39f)

amyeroberts commented:
@Rocketknight1 Awesome ! Am I right in saying the profiling shows this is faster now? 

You know the next question I'm going to ask is if it correctly handles when tokenizers have different splitting behaviour e.g. preprending with `##` :)

Rocketknight1 commented:
I'm working on that and _highly confident_ (not at all confident) that a solution can be found

Rocketknight1 committed:
Manual patching for BERT-like tokenizers

Rocketknight1 commented:
Quick update - `tokenizer.convert_tokens_to_string()` is actually a general method that does what we want, but it has the problem that it strips prefix spaces in tokenizer classes where prepending prefix spaces is the default, and I can't figure out how to make that stop! I think we can use it, but I might need a method where I always prepend with a fixed token before calling it or something to stop that behaviour.

Rocketknight1 committed:
Return a bool vector instead of a single True/False

Rocketknight1 committed:
Better comment

Rocketknight1 committed:
Better comment

Rocketknight1 commented:
@zucchini-nlp The condition now returns a per-sample vector correctly. Can I be lazy and ask you to add the test for it that @amyeroberts was requesting in #29116 here? If you're too busy, don't worry, I'll get to it!

Rocketknight1 commented:
@zucchini-nlp you can just add the test to this PR's branch instead!

Rocketknight1 committed:
Add tests from @zucchini-nlp

gante commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -155,3 +246,90 @@ def validate_stopping_criteria(stopping_criteria: StoppingCriteriaList, max_leng
     return new_stopping_criteria
+
+
+def _stop_string_get_matching_positions(
```
I'd add a comment explaining why we have two external functions, on each function (this one for testing purposes, the other one for caching purposes)

gante reviewed (approved):
🔥

amyeroberts reviewed (commented):
Thanks for iterating on this! 

I have two main reservations at the moment: 
* The logic is still extremely brittle - we essentially account for some tokenizers we've encountered but it doesn't generalise 
* The logic is pretty impenetrable. In fact, I'm really struggling to review this because the goals and intended behaviour of the functions isn't clear.

For 1) I'm OK to accept as a pragmatic approach. For 2) I'd like for there to be more explanation in the docstring about what should and shouldn't match - this should also be reflected in tests, which show explicitly when a string should be matched or not. At the moment, some of the tests e.g.  `test_stop_string_matching_positions` don't make it any clearer the desired behaviour, and is obfuscated by unnecessary conditional and looping logic.

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,174 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+        self.max_valid_end_lens = {
+            stop_string: max([len(val) for val in self.strings_to_end_lengths[stop_string].values()])
+            for stop_string in stop_strings
+        }
```
Done! (Can't merge automatically because of changes I made since then)

Rocketknight1 committed:
Amy's list creation nit

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,160 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+        validly appear in the stop strings. For each stop string, it returns a dictionary mapping tokens to a list of
+        valid positions, as well as a dictionary mapping tokens to a list of possible overlap lengths at the
+        end of the stop string."""
+        tok_list = list(self.vocab.keys())
```
Done!

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -126,6 +129,160 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
+        # The maximum length we need to consider is 1 token per character. Note that input_ids can also be
+        # *shorter* than the global max, and the code below should be ready for that
+        maximum_token_len = max([len(stop_string) for stop_string in self.stop_strings])
```
Done!

Rocketknight1 committed:
tok_list -> token_list

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -155,3 +246,90 @@ def validate_stopping_criteria(stopping_criteria: StoppingCriteriaList, max_leng
+    return token_valid_positions, token_end_overlaps
+
+
+@lru_cache(8)
```
I needed a number! 8 might be too high, but the cached values aren't especially large, either.

Rocketknight1 committed:
Push a big expanded docstring (should we put it somewhere else?)

Rocketknight1 committed:
Expand docstrings

Rocketknight1 committed:
Docstring fixups

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -155,3 +246,90 @@ def validate_stopping_criteria(stopping_criteria: StoppingCriteriaList, max_leng
+def _stop_string_get_matching_positions(
+    tok_list, tok_indices, stop_strings
+) -> Tuple[Dict[str, Dict[str, List[int]]], Dict[str, Dict[str, List[int]]]]:
+    """This function preprocesses stop strings and the tokenizer vocabulary to determine where tokens can
```
I wrote a real big docstring covering the entire algorithm!

Rocketknight1 replied:
(Maybe this should go somewhere else that isn't the docstring)

Rocketknight1 committed:
Rebase

Rocketknight1 committed:
make fixup

Rocketknight1 committed:
Make a properly general method for figuring out token strings

Rocketknight1 committed:
Fix naming throughout the functions

Rocketknight1 committed:
Move cache, refactor, fix tests

Rocketknight1 committed:
Add comment

Rocketknight1 committed:
Remove finished TODO

Rocketknight1 commented:
cc @amyeroberts this should be ready for re-review!

A quick summary of the changes:
- The old method of hardcoding the removal of prefixes like `##` is gone - it was way too brittle, as you mentioned. I have a new system [here](https://github.com/huggingface/transformers/pull/28932/files#diff-6e63ae0764aa864afd5bae6d512677b99b5240cb98cb210190482bdbb6a85906R298-R316) that I think should be robust across all of our tokenizers, unless there's a really weird one I'm missing.
- The cache system has been reworked, and the methods that were factored out of `StopStringCriteria` to allow them to be wrapped with `lru_cache()` have been moved back in. Instead, I use a manual module-level cache, which is keyed by vocabulary and stop strings. I do it this way because the preprocessing stage requires the `tokenizer`, and I didn't want to use that as a key. Therefore, the cache assumes that we won't have two tokenizers with wildly different preprocessing methods but identical vocabularies - this seems safe to me!
- Tests have been reworked for these two changes, and should be passing now
- There's a massive docstring explaining the workings of the class - I can move this somewhere else if it makes more sense there, since it might clog up the docs as-is!

Rocketknight1 committed:
Remove finished TODO

Rocketknight1 committed:
make fixup

amyeroberts commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -129,6 +136,351 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+    - ["stop"]  (overlap is "stop", overlap length == 4)
+    - ["st", "opera"]  (overlap is "op", overlap length == 2)
+    - ["sto", "pper"]  (overlap is "p", overlap length == 1)
+    - ["las", "topper"]  (overlap is "top", overlap length == 3)
```
What would happen if it's split over three tokens e.g. `["st", "o", "pper"]`?

Rocketknight1 replied:
Added an example!

amyeroberts commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -129,6 +136,351 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+    some i > 0. If you look at the positive examples above, you'll see the last token in all of them fulfills this
+    property:
+
+    - ["st", "op"] (overlap is "op", overlap length == 2)
+    - ["stop"]  (overlap is "stop", overlap length == 4)
+    - ["st", "opera"]  (overlap is "op", overlap length == 2)
+    - ["sto", "pper"]  (overlap is "p", overlap length == 1)
+    - ["las", "topper"]  (overlap is "top", overlap length == 3)
```
This part of the example is great. The rest of the description is quite wordy. I think it would be a lot clearer if the algorithm was walked through step by step with examples relating the strings to values calculated in the code e.g. `valid_positions`

Rocketknight1 replied:
I tried to shorten it and keep it example-focused, but I'm struggling to actually explain the algorithm in a small space!

amyeroberts commented on tests/generation/test_stopping_criteria.py:
```diff
@@ -124,3 +125,181 @@ def test_validate_stopping_criteria(self):
+        for i in range(len(false_strings)):
+            self.assertFalse(criteria(false_input_ids["input_ids"][i : i + 1], scores))
+
+    def test_stop_string_matching_positions(self):
```
For both this test and `test_stop_string_embedding_vecs`, the tests should be simplified: tests should contain as little if/else and looping logic as possible. They should be extremely clear and explicit. At the moment these tests require looking into the actual code to figure out if they're doing what's intended, whereas they should really form some kind of documentation - 

I'd suggest for each picking two stop words which satisfy different logic conditions and having each of their testing logic set out step-by-step. It's OK if code is repeated in tests

amyeroberts reviewed (commented):
Thanks for iterating on this and finding a super fast vectorization solution! 

I still think we need a bit of iteration here to make the functionality clearer - I'm struggling to review and make sure the logic is sound. The two biggest things that would help here are:
- Simplify the tests. If the tests are clear and explicit in what they're testing, I'll be more confident about the code logic :) 
- Simplify the docstring. Thank you for adding this! It really hepls having the logic documented out. It just needs another pass to make a bit terser and ideally show some clearer relations of strings to values calculated in the algo

The reason I'm being tough on this is that the code being as clear as possible makes sure others can add or fix, even if you're not available to answer questions about it

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -1,15 +1,22 @@
 
 
 logger = logging.get_logger(__name__)
+# We maintain a module-level cache of the embedding vectors for the stop string criterion
+# because they are slow to compute
+STOP_STRING_EMBEDDING_CACHE = OrderedDict()
```
I thought about it, but it's okay! The reason is that the cache is keyed by stop string and vocabulary, so the global cache could only cause a bug if two different models have identical vocabularies, but also handle things like `SPIECE_UNDERLINE` differently (if they have identical vocabularies **and** identical token handling, then the cached entry that works for one will also work for the other) This seems like a very unlikely combination of events to me!

amyeroberts replied:
I don't see why they'd need to have exactly the same vocabularies. Wouldn't there be clashes for any shared vocab keys?

Rocketknight1 replied:
The cache is keyed by `(token_list, token_indices, stop_strings)` in `clean_and_embed_tokens_with_cache`. "token list" and "token indices" here refer to the **entire** vocabulary of the tokenizer.

amyeroberts replied:
Ah, gotcha, in this case it is indeed very unlikely!

Rocketknight1 committed:
Update src/transformers/generation/stopping_criteria.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

Rocketknight1 commented on src/transformers/generation/stopping_criteria.py:
```diff
@@ -129,6 +136,351 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwa
+    - ["stop"]
+    - ["st", "opera"]
+    - ["sto", "opper"]
+    - ["las", "topper"]
```
That would also be valid! I'll add three-token examples too

Rocketknight1 committed:
Update and shorten docstring

Rocketknight1 committed:
Update tests to be shorter/clearer and test specific cases

amyeroberts commented on tests/generation/test_stopping_criteria.py:
```diff
@@ -124,3 +125,134 @@ def test_validate_stopping_criteria(self):
+        self.assertEqual(valid_positions, {"s": [3], "last": [2]})
+        self.assertEqual(end_overlaps, {"top": [3], "topper": [3], "p": [1]})
+
+    def test_stop_string_embedding_vecs(self):
```
❤️

amyeroberts reviewed (approved):
Thanks for all the work adding and iterating on this!
